[Typechecker] Allow bridging of Unmanaged to Objective-C for throwing functions

theblixguy opened the pull request:
Previously, we would emit an error diagnostic if a throwing `@objc` function returned `Unmanaged<T>`. This is now allowed when `T` is representable in Objective-C.

For example:

```swift
class Foo {}

@objc protocol Bar {
  func throwingMethod1() throws -> Unmanaged<CFArray> // ok
  func throwingMethod2() throws -> Unmanaged<Foo> // error
}
```
  

Resolves [SR-9035](https://bugs.swift.org/browse/SR-9035).

theblixguy commented:
cc @jrose-apple

theblixguy committed:
[typechecker] allow a throwing @objc function to return Unmanaged<T>

jrose-apple commented:
Thanks for picking this up! I'm worried that we won't necessarily compile this correctly, though, so can you add an execution test where a Swift thing is being called from Objective-C and the Objective-C side handles both the success and failure cases?

jrose-apple commented:
(Alternately, an IRGen test, but only if you want to practice writing IRGen tests.)

theblixguy committed:
[test] Adds an IRGen test

jrose-apple commented on test/IRGen/unmanaged_objc_throw_func.swift:
```diff
@@ -0,0 +1,27 @@
+    // CHECK-NEXT: call void @swift_bridgeObjectRelease(%swift.bridge* %4) #2
+    // CHECK-NEXT: %9 = bitcast %TSo10CFArrayRefa* %8 to i8*
+    // CHECK-NEXT: call void @llvm.objc.release(i8* %9)
+    // CHECK-NEXT: ret %TSo10CFArrayRefa* %8
```
This is good, but what we really want to check is the `@objc` thunk for that wraps this method, to make sure it returns the value in the success case and `NULL` in the failure case.

theblixguy replied:
Oh I see, I have updated the test file now!

theblixguy committed:
[test] Updates IRGen test file

theblixguy committed:
[test] remove -primary-file flag as it's not needed

jrose-apple commented on test/IRGen/unmanaged_objc_throw_func.swift:
```diff
@@ -0,0 +1,70 @@
+// RUN: %target-swift-frontend -emit-ir %s | %FileCheck %s
```
Should mark this test as `REQUIRES: objc_interop`, since it's using the real Foundation and CF.

jrose-apple commented on test/IRGen/unmanaged_objc_throw_func.swift:
```diff
@@ -0,0 +1,70 @@
+  func returnUnmanagedCFArray() throws -> Unmanaged<CFArray>
+}
+
+// CHECK: define hidden swiftcc %TSo10CFArrayRefa* @"$s25unmanaged_objc_throw_func9SR_9035_CC22returnUnmanagedCFArrays0G0VySo0H3RefaGyKF"(%T25unmanaged_objc_throw_func9SR_9035_CC* swiftself, %swift.error** noalias nocapture swifterror dereferenceable(8)) #0 {
```
Using `CHECK-LABEL` instead of `CHECK` here helps FileCheck divide the file up into separate sections, which keeps it from exiting early when you're checking multiple things in one file.

jrose-apple replied:
Also, if more tests get added to this file, that `#0` (the function attributes) might change. Since they're not important, it's probably best to use a regex match here: `#{{[0-9]+}}`.

jrose-apple commented on test/IRGen/unmanaged_objc_throw_func.swift:
```diff
@@ -0,0 +1,70 @@
+// CHECK: define hidden swiftcc %TSo10CFArrayRefa* @"$s25unmanaged_objc_throw_func9SR_9035_CC22returnUnmanagedCFArrays0G0VySo0H3RefaGyKF"(%T25unmanaged_objc_throw_func9SR_9035_CC* swiftself, %swift.error** noalias nocapture swifterror dereferenceable(8)) #0 {
+@objc class SR_9035_C: NSObject, SR_9035_P {
+  func returnUnmanagedCFArray() throws -> Unmanaged<CFArray> {
+    // CHECK: %3 = call swiftcc { %swift.bridge*, i8* } @"$ss27_allocateUninitializedArrayySayxG_BptBwlF"(i64 1, %swift.type* @"$sSiN")
```
The SSA value numbers (`%3` et al) are pretty closely tied to the particular IR that gets emitted today. To avoid that, you can use a variable pattern: `%[[ALLOCATION:.+]] = call …`. Then in later expressions, you can just use `%[[ALLOCATION]]` to refer to `%3`, and that way it'll be easier to update if an instruction is added or deleted later.

(These features are all in https://llvm.org/docs/CommandGuide/FileCheck.html if you want to take a look later.)

jrose-apple commented on test/IRGen/unmanaged_objc_throw_func.swift:
```diff
@@ -0,0 +1,70 @@
+  } 
+}
+
+// CHECK: %14 = call swiftcc %TSo10CFArrayRefa* @"$s25unmanaged_objc_throw_func9SR_9035_CC22returnUnmanagedCFArrays0G0VySo0H3RefaGyKF"(%T25unmanaged_objc_throw_func9SR_9035_CC* swiftself %4, %swift.error** noalias nocapture swifterror dereferenceable(8) %swifterror) #6
```
Same as the above function advice…

jrose-apple commented on test/IRGen/unmanaged_objc_throw_func.swift:
```diff
@@ -0,0 +1,70 @@
+// CHECK: %14 = call swiftcc %TSo10CFArrayRefa* @"$s25unmanaged_objc_throw_func9SR_9035_CC22returnUnmanagedCFArrays0G0VySo0H3RefaGyKF"(%T25unmanaged_objc_throw_func9SR_9035_CC* swiftself %4, %swift.error** noalias nocapture swifterror dereferenceable(8) %swifterror) #6
+// CHECK-NEXT: %15 = load %swift.error*, %swift.error** %swifterror, align 8
+// CHECK-NEXT: %16 = icmp ne %swift.error* %15, null
+// CHECK-NEXT: br i1 %16, label %20, label %17
```
…and same advice about patterns here for the labels to jump to.

jrose-apple commented on test/IRGen/unmanaged_objc_throw_func.swift:
```diff
@@ -0,0 +1,70 @@
+// CHECK-NEXT: %16 = icmp ne %swift.error* %15, null
+// CHECK-NEXT: br i1 %16, label %20, label %17
+
+// CHECK: ; <label>:17:                                     ; preds = %entry
```
You can also leave out things that aren't so relevant to what's being tested, like the "preds" line here.

jrose-apple commented on test/IRGen/unmanaged_objc_throw_func.swift:
```diff
@@ -0,0 +1,70 @@
+// CHECK-NEXT: %39 = bitcast %T25unmanaged_objc_throw_func9SR_9035_CC* %4 to i8*
+// CHECK-NEXT: call void @llvm.objc.release(i8* %39)
+// CHECK-NEXT: %40 = inttoptr i64 %38 to %struct.__CFArray**
+// CHECK-NEXT: ret %struct.__CFArray** %40
```
…and here's why we were testing it: this doesn't look like the right type. I think the *values* are all correct, but that type should have one fewer `*`, don't you think?

(Although maybe that's an existing bug. Worth checking against the behavior for plain `CFArray`.)

theblixguy replied:
Hmm if I return `arr` instead of `Unmanaged.passUnretained(arr)` then we get

```
%43 = inttoptr i64 %42 to %struct.__CFArray*
ret %struct.__CFArray* %43
```

jrose-apple replied:
:-/ I guess we have a bug, then. How does the SIL look? Do we have extra indirection there?

theblixguy replied:
For `Unmanaged<CFArray>`, this is what the last couple of instructions look:

```asm
%14 = unchecked_ref_cast %13 : $NSArray to $CFArray // user: %15
%15 = unchecked_ref_cast %14 : $CFArray to $CFArray // users: %18, %16
debug_value %15 : $CFArray, let, name "arr"     // id: %16
release_value %6 : $Array<Int>                  // id: %17
%18 = ref_to_unmanaged %15 : $CFArray to $@sil_unmanaged CFArray // user: %19
%19 = struct $Unmanaged<CFArray> (%18 : $@sil_unmanaged CFArray) // user: %20
return %19 : $Unmanaged<CFArray>                // id: %20
```

For just `CFArray`:

```asm
%14 = unchecked_ref_cast %13 : $NSArray to $CFArray // user: %15
%15 = unchecked_ref_cast %14 : $CFArray to $CFArray // users: %18, %16
debug_value %15 : $CFArray, let, name "arr"     // id: %16
release_value %6 : $Array<Int>                  // id: %17
return %15 : $CFArray
```

jrose-apple replied:
Ah, as before it's not this function that matters, but the one that bridges to ObjC.

theblixguy replied:
For `Unmanaged<CFArray>`, the thunk ends with
 
```asm
return %25 : $Optional<Unmanaged<CFArray>>
```

For `CFArray`

```asm
return %25 : $Optional<CFArray>
```

jrose-apple replied:
Hmph. Both of those seem reasonable, so we must have a bug in IRGen. Drat! That means we can't just take this change.

theblixguy replied:
Oh! It seems to compile and run correctly however (at least in Swift). Anyway, I could file a JIRA for this if you want?

jrose-apple replied:
I think it's still going to block *this* JIRA. It may not be causing problems now, but it could in the future (or possibly in a slightly more complicated test case).

You could test it a little harder in Swift by using `dynamic`, which will force the compiler to use `objc_msgSend` even for a Swift-defined method.

theblixguy replied:
Hmm. It works even after I add `dynamic` to the method. What would you suggest the best thing to do right now would be? Do we let this stay blocked until someone (maybe John?) takes a look at it or do we write more tests to ensure it behaves correctly?

jrose-apple replied:
At this point I think someone has to look into why it's behaving incorrectly, _or_ show that it's not a regression.

You know, one thing we haven't tried is seeing what a normal `Optional<Unmanaged<CFArray>>` return looks like.

theblixguy replied:
IRGen:

```asm
%9 = ptrtoint %TSo10CFArrayRefa* %8 to i64
%10 = bitcast %TSo10CFArrayRefa* %8 to i8*
call void @llvm.objc.release(i8* %10)
ret i64 %9
```

SIL:

```
%18 = ref_to_unmanaged %15 : $CFArray to $@sil_unmanaged CFArray // user: %19
%19 = struct $Unmanaged<CFArray> (%18 : $@sil_unmanaged CFArray) // user: %20
%20 = enum $Optional<Unmanaged<CFArray>>, #Optional.some!enumelt.1, %19 : $Unmanaged<CFArray> // user: %22
strong_release %15 : $CFArray                   // id: %21
return %20 : $Optional<Unmanaged<CFArray>>
```

jrose-apple replied:
\*shakes head\* Then yeah, something's going wrong specifically in the throwing case.

theblixguy replied:
That was without using `@objc`. If I use `@objc`, but remove the `throws`, then I get the same IR/SIL:

```
%13 = inttoptr i64 %11 to %struct.__CFArray**
ret %struct.__CFArray** %13
```

So it seems like the we get the same code with or without the `throws`. But if I remove `@objc`, then we get the code I mentioned in my previous comment.

jrose-apple replied:
Bizarre, but okay, I guess we can treat it as a separate bug. (And this is a better reproducer for that bug.) Thanks for looking into it!

jrose-apple reviewed (commented):
Thanks! I have a bunch of FileCheck tips now on how to make this test more robust.

theblixguy committed:
[test] Update IRGen test with patterns, etc

theblixguy commented:
Thanks Jordan, I have updated the test file with your suggestions.

jrose-apple reviewed (commented):
Heh, okay. More semantic names are nicer for maintainers, but <del>this is good enough to run tests!</del> …except for the return type thing.

jrose-apple commented:
@swift-ci Please test

theblixguy commented:
Hmm fails on iOS simulator: 

```
/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/IRGen/unmanaged_objc_throw_func.swift:10:17: error: CHECK-LABEL: expected string not found in input
18:58:50 // CHECK-LABEL: define hidden swiftcc %TSo10CFArrayRefa* @"$s25unmanaged_objc_throw_func9SR_9035_CC22returnUnmanagedCFArrays0G0VySo0H3RefaGyKF"(%T25unmanaged_objc_throw_func9SR_9035_CC* swiftself, %swift.error** noalias nocapture swifterror dereferenceable(8)) #{{[0-9]+}} {
18:58:50                 ^
18:58:50 <stdin>:1:1: note: scanning from here
18:58:50 ; ModuleID = '-'
18:58:50 ^
```

could just remove the `CHECK-LABEL` as it's not important what the function is called.

jrose-apple commented:
It matters which function it is, though. I suspect the difference is going to be in all those attributes of the error parameter…at the very least the pointer size is different. It's probably worth testing locally with a 32-bit simulator (watchOS is easiest to get working) to see where else the CHECK lines are assuming pointer sizes!

jrose-apple commented:
This turned out to be a lot more than I thought it would be.

theblixguy commented:
How do I test with a simulator? Is it just a matter of adding “REQUIRES: ios” to the test file or is there a special flag I need to pass to the build script?

jrose-apple commented:
Ah, check out docs/Testing.md and the sample `lit.py` invocation. The test directory you pick determines which target you're testing.

theblixguy committed:
Merge branch 'master' into fix/SR-9035

theblixguy committed:
[test] Update IRGen test

theblixguy commented:
Sorry for the delay. `lit` didn't work for me, but I was able to get the test running on the watchOS simulator using `cmake`. 

So basically, we have a bunch of `i64`s and on 32-bit devices, we want to ensure they're `i32`s instead, so I have added a pattern to do that. Same with alignment (sometimes we have `4`, sometimes `8`).

Additionally, the `CHECK-LABEL` was failing because the string didn't contain `swifterror`. I have added a pattern to make it optional, however it's not working. Is there something wrong with the pattern `{{(swifterror?)}}`? It should work as it's valid regex.

jrose-apple commented:
I think we don't have a special swifterror convention on 32-bit platforms either. Something like that.

@swift-ci Please test

theblixguy commented:
Oh! Seems like the macOS test just disappeared! 

Anyway, if it fails then we either need add “REQUIRES: CPU=x86_64” and just not run this in 32-bit or add a “CHECK-LABEL” for each architecture (`CHECK-LABEL-i386`, etc).

jrose-apple commented:
Hrm.

@swift-ci Please test macOS

jrose-apple commented:
I suggest cutting down the CHECK-LABEL line to just match the full name of the function. That should be sufficiently unique, and it's not likely that the parameters are being emitted wrong.

```
CHECK-LABEL: define hidden swiftcc %TSo10CFArrayRefa* @"$s25unmanaged_objc_throw_func9SR_9035_CC22returnUnmanagedCFArrays0G0VySo0H3RefaGyKF"
```

theblixguy committed:
[Test] Trim CHECK-LABEL

theblixguy commented:
Yeah, I've done that now!

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
@swift-ci Please smoke test Linux

theblixguy commented:
Tests have passed! Can I cherry-pick this to 5.1?

jrose-apple commented:
This is a pretty contained change, and it only enables more code than before. Yes, I think that's okay!
